js.go

```diff
@@ -1250,6 +1369,128 @@ func (jsi *jsSub) trackSequences(reply string) {
 	jsi.mu.Unlock()
 }
 
+// Check to make sure messages are arriving in order.
+// Returns true if the sub had to be replaced. Will cause upper layers to return.
+// Lock should be held.
+func (sub *Subscription) checkOrderedMsgs(m *Msg) bool {
+	// Ignore msgs with no reply like HBs and flowcontrol, they are handled elsewhere.
+	if m.Reply == _EMPTY_ || sub.jsi == nil || isControlMessage(m) {
+		return false
+	}
+
+	// Normal message here.
+	tokens, err := getMetadataFields(m.Reply)
+	if err != nil {
+		return false
+	}
+	sseq, dseq := uint64(parseNum(tokens[5])), uint64(parseNum(tokens[6]))
+
+	jsi := sub.jsi
+	jsi.mu.Lock()
+	if dseq != jsi.dseq {
+		jsi.mu.Unlock()
+		sub.resetOrderedConsumer(jsi.sseq + 1)
```

Comment:
Should you compute `jsi.sseq+1` while under the jsi lock?

Reply:
We can grab the value yes. Will update.